feat: Adds support for region & customer_endpoint_dns_name in privatelink

EspenAlbert opened the pull request:
## Description

Adds support for `region` & `customer_endpoint_dns_name` in `privatelinkendpointservicedatafederationonlinearchive`.
Also, enables editing of the `comment` attribute.

Link to any related issue(s): CLOUDP-229354
GH Issue: https://github.com/mongodb/terraform-provider-mongodbatlas/issues/1878


## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

## Further comments

Minimal setup for testing manually:
```hcl
variable "project_id" {
  type = string
}

variable "vpc_id" {
  type = string 
}

variable "subnet_ids" {
  type = list(string)
}
variable "security_group_ids" {
  type = list(string)
}

resource "aws_vpc_endpoint" "vpce_east" {
  vpc_id             = var.vpc_id
  service_name       = "com.amazonaws.vpce.us-east-1.vpce-svc-0a7247db33497082e" #see screenshot below
  vpc_endpoint_type  = "Interface"
  subnet_ids         = var.subnet_ids
  security_group_ids = var.security_group_ids
}

resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archive" "test" {
  project_id    = var.project_id
  endpoint_id   = aws_vpc_endpoint.vpce_east.id
  provider_name = "AWS"

  comment       = "Terraform Acceptance Test Manual"
  region        = "US_EAST_1"
  customer_endpoint_dns_name = aws_vpc_endpoint.vpce_east.dns_entry[0].dns_name
}
output "vpc_endpoint_id" {
  value = aws_vpc_endpoint.vpce_east.id
}
```

How to find the `service_name`:

![image](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/9f753aea-a593-4fa4-ac7f-e021dcccb845)

After creating the resource in the Atlas UI:

![image](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/128c3519-c79f-4d17-bd76-b507ae3fa351)

Running CI tests locally: 

```shell
export MONGODB_ATLAS_LAST_VERSION="1.15.1"
export MONGODB_ATLAS_PRIVATE_ENDPOINT_ID={OUTPUTS.vpc_endpoint_id}
go test -timeout 2400s -run ^(TestAccNetworkPrivatelinkEndpointServiceDataFederationOnlineArchivesDSPlural_basic|TestAccMigrationNetworkPrivatelinkEndpointServiceDataFederationOnlineArchive_basic|TestAccNetworkPrivatelinkEndpointServiceDataFederationOnlineArchive_basic|TestAccNetworkPrivatelinkEndpointServiceDataFederationOnlineArchive_updateComment|TestAccNetworkPrivatelinkEndpointServiceDataFederationOnlineArchive_basicWithRegionDnsName|TestAccNetworkPrivatelinkEndpointServiceDataFederationOnlineArchiveDS_basic)$ github.com/mongodb/terraform-provider-mongodbatlas/internal/service/privatelinkendpointservicedatafederationonlinearchive -count=1 -v
```

![image](https://github.com/mongodb/terraform-provider-mongodbatlas/assets/6596097/d4bf88a0-73f9-4363-ae0e-b1cd700e4649)

EspenAlbert committed:
feat: add support for region & customer_endpoint_dns_name in privatelink

EspenAlbert committed:
refactor: move StringIsUppercase to validate package

EspenAlbert committed:
refactor: use a shared function for populating comment, type, region ,etc.

EspenAlbert committed:
docs: add documentation for region and custom_endpoint_dns_name

EspenAlbert committed:
docs: fix error message text typo

EspenAlbert commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/resource_privatelink_endpoint_service_data_federation_online_archive.go:
```diff
@@ -53,6 +54,17 @@ func Resource() *schema.Resource {
+			"region": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				ForceNew:         true,
```
Using `ForceNew` here since the Update function is not implemented, but I'm uncertain if it is correct.
On this resource, I think only `comment` should be `ForceNew: false`, (but it is currently `true`).
The [api](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/v2/#tag/Data-Federation/operation/createDataFederationPrivateEndpoint)  don't expose a `PUT` endpoint but mentions updating when only `comment` has changed.

Is it worth it to implement the Update function?

maastha replied:
I would just try this out via the API once and if updating only comment is supported correctly (i.e. no impact to remaining attributes) then I don't see why we cannot support update for comment

EspenAlbert replied:
Works. Thanks to @oarbusi for the help 🙏

EspenAlbert committed:
fix: add extra fields to data source

maastha commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/resource_privatelink_endpoint_service_data_federation_online_archive.go:
```diff
@@ -200,5 +196,36 @@ func newPrivateNetworkEndpointIDEntry(d *schema.ResourceData) *admin.PrivateNetw
 }
+
+func populateResourceData(d *schema.ResourceData, privateEndpoint *admin.PrivateNetworkEndpointIdEntry) error {
+	if err := d.Set("comment", privateEndpoint.GetComment()); err != nil {
```
I'd include setting project_id & endpoint_id here

EspenAlbert replied:
Done 👍

maastha commented on website/docs/r/privatelink_endpoint_service_data_federation_online_archive.html.markdown:
```diff
@@ -33,13 +33,15 @@ resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archi
 * `endpoint_id` (Required) - Unique 22-character alphanumeric string that identifies the private endpoint. See [Atlas Data Lake supports Amazon Web Services private endpoints using the AWS PrivateLink feature](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Data-Federation/operation/createDataFederationPrivateEndpoint:~:text=Atlas%20Data%20Lake%20supports%20Amazon%20Web%20Ser ...
 * `provider_name` (Required) - Human-readable label that identifies the cloud service provider. 
 * `timeouts`- (Optional) The duration of time to wait for Private Endpoint Service to be created or deleted. The timeout value is definded by a signed sequence of decimal numbers with an time unit suffix such as: `1h45m`, `300s`, `10m`, .... The valid time units are:  `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`. The default timeout for Private Endpoint create & delete is `2h`. Learn more about time ...
+* `region` - (Optional) Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft- ...
```
there would be two similar doc under `website/docs/d` for data sources that need to be updated as well

EspenAlbert replied:
Added. Wasn't sure whether or not to include (Optional) in the data sources

maastha replied:
depends, is this attribute required to be provided by the user when calling this data source? what happens if the user doesn't provide it? how is the API call made then?
Did you try using the data source to get a better idea around usage?

EspenAlbert replied:
It is never used when reading.
Maybe better to mark is as `Computed`?

maastha replied:
yes should be computed in data source, I added a comment on the data source file itself so there's no confusion, let's resolve this one

EspenAlbert committed:
feat: support updating comment

EspenAlbert committed:
test: add a test for region and customer_endpoint_dns_name

EspenAlbert committed:
test: refactor comment to global var

EspenAlbert committed:
refactor: add projectID & endpointID to populateResourceData

EspenAlbert committed:
test: add region and customerEndpointDNSName to data source test

EspenAlbert committed:
docs: add new attributes to data sources documentation

maastha commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/resource_privatelink_endpoint_service_data_federation_online_archive_test.go:
```diff
@@ -92,13 +168,33 @@ func checkExists(resourceName string) resource.TestCheckFunc {
+	`, projectID, endpointID, comment)
+}
+
+func asCustomerEndpointDNSName(endpointID string) string {
```
```suggestion
func customerEndpointDNSName(endpointID string) string {
```

maastha replied:
or toCustomerEndpointDNSName or something?

EspenAlbert replied:
Renamed to `toCustomerEndpointDNSName`, as I'm using customerEndpointDNSName as the variable name

maastha commented on website/docs/r/privatelink_endpoint_service_data_federation_online_archive.html.markdown:
```diff
@@ -33,13 +33,15 @@ resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archi
 * `timeouts`- (Optional) The duration of time to wait for Private Endpoint Service to be created or deleted. The timeout value is definded by a signed sequence of decimal numbers with an time unit suffix such as: `1h45m`, `300s`, `10m`, .... The valid time units are:  `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`. The default timeout for Private Endpoint create & delete is `2h`. Learn more about time ...
+* `region` - (Optional) Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft- ...
+* `customer_endpoint_dns_name` - (Optional) Human-readable label to identify customer's VPC endpoint DNS name. Must follow regex: `vpce-[a-z0-9]{17}-[a-z0-9]{8}.vpce-svc-[a-z0-9]{17}.(.*).vpce.amazonaws.com`
+* `comment` - (Optional) Human-readable string to associate with this private endpoint.
```
did we confirm if it's okay to allow users to update comment?

AgustinBettati replied:
if we support the update for the comment, might be worth add a note at the start of the docs mentioning that the update operation is limited to the comment attribute.

EspenAlbert replied:
I added a note now. You can resolve if it looks ok

EspenAlbert replied:
Changes made in 7c55d85 ✅

EspenAlbert committed:
refactor: rename according to review

maastha commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/data_source_privatelink_endpoint_service_data_federation_online_archives.go:
```diff
@@ -43,6 +43,14 @@ func PluralDataSource() *schema.Resource {
 						},
+						"region": {
+							Type:     schema.TypeString,
+							Optional: true,
```
if these cannot be set by the user, these should be `Computed` in both data sources

EspenAlbert replied:
Fixed in 1a0ba3a and 80c0762

EspenAlbert committed:
fix: make data source attributes computed

EspenAlbert committed:
fix: make single data source attributes computed

lantoli commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/data_source_privatelink_endpoint_service_data_federation_online_archive_test.go:
```diff
@@ -15,6 +15,7 @@ var (
 func TestAccNetworkPrivatelinkEndpointServiceDataFederationOnlineArchiveDS_basic(t *testing.T) {
 	// Skip because private endpoints are deleted daily from dev environment
 	acc.SkipTestForCI(t)
+	customerEndpointDNSName := toCustomerEndpointDNSName(endpointID)
```
as this test is skipped in CI, were you able to run it in local?

EspenAlbert replied:
Yes. I was able to run all the tests.  I included a [screenshot](https://private-user-images.githubusercontent.com/6596097/310139387-d4bf88a0-73f9-4363-ae0e-b1cd700e4649.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk2NjQ4MDUsIm5iZiI6MTcwOTY2NDUwNSwicGF0aCI6Ii82NTk2MDk3LzMxMDEzOTM4Ny1kNGJmODhhMC03M2Y5LTQzNjMtYWUwZS1iMWNkNzAwZTQ2NDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDVUMTg0ODI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGJlNTJkOGE2N2VlNzMxODM2ODY0NmI1MmZlY2Q2M2Q0M2U0ZTQ5NDQ1YzU4ZTBjMzlkZWU3OWM1MzE5YjU2ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.parxFFA78zRfq8sSu2RTOOYpDnzkYE1rDIpJR0JXvxo) in the bottom of the description of this PR

lantoli commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/resource_privatelink_endpoint_service_data_federation_online_archive_test.go:
```diff
@@ -27,12 +29,86 @@ func TestAccNetworkPrivatelinkEndpointServiceDataFederationOnlineArchive_basic(t
+}
+func TestAccNetworkPrivatelinkEndpointServiceDataFederationOnlineArchive_updateComment(t *testing.T) {
+	// Skip because private endpoints are deleted daily from dev environment
+	acc.SkipTestForCI(t)
```
similar comment, were you able to run all skipped tests in local?

marcosuma replied:
@lantoli curious to know your opinion: is there an opportunity to actually convert this to run in CI? not sure how big the work is

marcosuma replied:
wondering if this is something that Espen could do it (not advocating for it)

andreaangiolillo replied:
we used to have them enabled in cloud-dev. We disabled them in https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1929 because cloud-dev started to delete the private endpoint every day similar to the cluster.

lantoli replied:
for example we could reuse a project that has cleanup skipped

EspenAlbert replied:
All that would be required AFAIK is having access to a `MONGODB_ATLAS_PROJECT_ID`, `MONGODB_ATLAS_ORG_ID`, `MONGODB_ATLAS_PRIVATE_ENDPOINT_ID`, and `MONGODB_ATLAS_PRIVATE_ENDPOINT_DNS_NAME`. The VPC Endpoint itself is not deleted in the AWS account. 
We'll address that in the test project :)

marcosuma commented on internal/service/advancedcluster/model_advanced_cluster.go:
```diff
@@ -247,21 +246,6 @@ func SchemaAdvancedConfig() *schema.Schema {
 	}
 }
 
-func StringIsUppercase() schema.SchemaValidateDiagFunc {
```
nice that you moved this!

marcosuma commented on website/docs/r/privatelink_endpoint_service_data_federation_online_archive.html.markdown:
```diff
@@ -33,13 +33,15 @@ resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archi
 * `provider_name` (Required) - Human-readable label that identifies the cloud service provider. 
 * `timeouts`- (Optional) The duration of time to wait for Private Endpoint Service to be created or deleted. The timeout value is definded by a signed sequence of decimal numbers with an time unit suffix such as: `1h45m`, `300s`, `10m`, .... The valid time units are:  `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`. The default timeout for Private Endpoint create & delete is `2h`. Learn more about time ...
+* `region` - (Optional) Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft- ...
+* `customer_endpoint_dns_name` - (Optional) Human-readable label to identify customer's VPC endpoint DNS name. Must follow regex: `vpce-[a-z0-9]{17}-[a-z0-9]{8}.vpce-svc-[a-z0-9]{17}.(.*).vpce.amazonaws.com`
```
is this regex imposed by us or from the backend? if so, could we reference the official documentation where this is validated as well?

EspenAlbert replied:
Yes, I learned this the "hard" way 😅 
I had to dig into the code to find it. The [API docs](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/v2/#tag/Data-Federation/operation/createDataFederationPrivateEndpoint)  didn't say anything.
I messaged Zuhair about it, but I didn't get a reply yet

Zuhairahmed replied:
Made a few suggestions directly in this PR. Cleaned up example to make sure reader is aware of appriorapate formats for `customer_endpoint_dns_name` and `endpoint_id`. Also updated docs definitions to be more inline with Admin API.

EspenAlbert replied:
Changes made in 7c55d85 ✅

AgustinBettati commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/resource_privatelink_endpoint_service_data_federation_online_archive_test.go:
```diff
@@ -92,13 +168,33 @@ func checkExists(resourceName string) resource.TestCheckFunc {
+	// Found in `.AWS.dev.us-east-1`:  https://github.com/10gen/mms/blob/85ec3df92711014b17643c05a61f5c580786556c/server/conf/data-lake-endpoint-services.json
+	serviceName := "vpce-svc-0a7247db33497082e"
+	return fmt.Sprintf("%s-8charsra.%s.us-east-1.vpce.amazonaws.com", endpointID, serviceName)
+}
```
Nice investigation here! Curious about the `s-8charsra` portion defined in the result, was this value specific to your setup for running the tests?

EspenAlbert replied:
This should be a random alphanumeric 8 character string created by aws.
Therefore, I created the `8charsra` (`char=characters` and `ra=random` ) 😅. I'll add a comment

AgustinBettati replied:
just to be sure I understand, if its created by aws would this have to be modified each time it is run? If that is the case I would favor extracting to an env variable to it is clear what needs to be adjusted (or provided externally). Otherwise happy to leave as is.

EspenAlbert replied:
You are right. Although the test was passing with any random value (I don't think the API checks that it matches the vpc endpoint dns name), this might not work in practice.
Using an env variable also enables us to run this test in different AWS region/environments where serviceName is different.
Thank you for checking this 😊

Changed in 09c4b80

AgustinBettati commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/resource_privatelink_endpoint_service_data_federation_online_archive.go:
```diff
@@ -49,6 +52,17 @@ func Resource() *schema.Resource {
 			"comment": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
```
nit: would remove the property if it has the default value of false

AgustinBettati commented on website/docs/d/privatelink_endpoint_service_data_federation_online_archive.markdown:
```diff
@@ -46,6 +46,8 @@ In addition to all arguments above, the following attributes are exported:
 * `type` - Human-readable label that identifies the resource type associated with this private endpoint.
```
I saw the renaming of *.markdown to *.html.markdown for the resource doc file, could we do the same for the data source files? Just for consistency with other doc files

maastha commented on website/docs/d/privatelink_endpoint_service_data_federation_online_archive.markdown:
```diff
@@ -46,6 +46,8 @@ In addition to all arguments above, the following attributes are exported:
 * `type` - Human-readable label that identifies the resource type associated with this private endpoint.
 * `comment` - Human-readable string to associate with this private endpoint.
 * `provider_name` - Human-readable label that identifies the cloud service provider. 
+* `region` - Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/).
```
could we also update the example for this resource with these new attributes and make sure it's working as expected?

EspenAlbert replied:
I updated the example in the resource documentation. Can you have a 2nd look and see if it is ok, and then I can also update the documentation of the data sources?

EspenAlbert committed:
refactor: apply suggestions from PR

EspenAlbert committed:
docs: add new arguments to example and note about `comment` being the only "updatable" argument

lantoli commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/resource_privatelink_endpoint_service_data_federation_online_archive_test.go:
```diff
@@ -13,9 +13,11 @@ import (
+	resourceName       = "mongodbatlas_privatelink_endpoint_service_data_federation_online_archive.test"
+	projectID          = os.Getenv("MONGODB_ATLAS_PROJECT_ID")
+	endpointID         = os.Getenv("MONGODB_ATLAS_PRIVATE_ENDPOINT_ID")
+	defaultComment     = "Terraform Acceptance Test"
```
knitpit: not sure why we say "default..." instead of just comment or region / atlasRegion

EspenAlbert replied:
changed in f4bd5e4 ✅

lantoli commented on internal/service/privatelinkendpointservicedatafederationonlinearchive/resource_privatelink_endpoint_service_data_federation_online_archive_test.go:
```diff
@@ -27,12 +29,86 @@ func TestAccNetworkPrivatelinkEndpointServiceDataFederationOnlineArchive_basic(t
+				),
+			},
+			{
+				Config: resourceConfigBasic(projectID, endpointID, "Terraform Acceptance Test Updated"),
```
extract constant as being used in a couple of times

EspenAlbert replied:
changed in f4bd5e4 ✅

lantoli reviewed (approved):
LGTM, just a couple of minor comments

EspenAlbert committed:
refactor: apply naming suggestions

Zuhairahmed commented on website/docs/r/privatelink_endpoint_service_data_federation_online_archive.html.markdown:
```diff
@@ -25,21 +28,26 @@ resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archi
 * `endpoint_id` (Required) - Unique 22-character alphanumeric string that identifies the private endpoint. See [Atlas Data Lake supports Amazon Web Services private endpoints using the AWS PrivateLink feature](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Data-Federation/operation/createDataFederationPrivateEndpoint:~:text=Atlas%20Data%20Lake%20supports%20Amazon%20Web%20Ser ...
 * `provider_name` (Required) - Human-readable label that identifies the cloud service provider. 
 * `timeouts`- (Optional) The duration of time to wait for Private Endpoint Service to be created or deleted. The timeout value is definded by a signed sequence of decimal numbers with an time unit suffix such as: `1h45m`, `300s`, `10m`, .... The valid time units are:  `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`. The default timeout for Private Endpoint create & delete is `2h`. Learn more about time ...
+* `region` - (Optional) Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft- ...
```
```suggestion
* `region` - (Optional) Human-readable label to identify the region of VPC endpoint.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/).
```

Zuhairahmed commented on website/docs/r/privatelink_endpoint_service_data_federation_online_archive.html.markdown:
```diff
@@ -25,21 +28,26 @@ resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archi
 * `provider_name` (Required) - Human-readable label that identifies the cloud service provider. 
 * `timeouts`- (Optional) The duration of time to wait for Private Endpoint Service to be created or deleted. The timeout value is definded by a signed sequence of decimal numbers with an time unit suffix such as: `1h45m`, `300s`, `10m`, .... The valid time units are:  `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`. The default timeout for Private Endpoint create & delete is `2h`. Learn more about time ...
+* `region` - (Optional) Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft- ...
+* `customer_endpoint_dns_name` - (Optional) Human-readable label to identify customer's VPC endpoint DNS name. Must follow regex: `vpce-[a-z0-9]{17}-[a-z0-9]{8}.vpce-svc-[a-z0-9]{17}.(.*).vpce.amazonaws.com`
```
```suggestion
* `customer_endpoint_dns_name` - (Optional) Human-readable label to identify VPC endpoint DNS name. Must follow regex: `vpce-[a-z0-9]{17}-[a-z0-9]{8}.vpce-svc-[a-z0-9]{17}.(.*).vpce.amazonaws.com`
```

Zuhairahmed commented on website/docs/r/privatelink_endpoint_service_data_federation_online_archive.html.markdown:
```diff
@@ -25,21 +28,26 @@ resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archi
   provider_name = "AWS"
   comment = "Test"
+  region        = "US_EAST_1"
+  customer_endpoint_dns_name = aws_vpc_endpoint.<AWS-RESOURCE_NAME>.dns_entry[0].dns_name
```
```suggestion
  customer_endpoint_dns_name = "vpce-046cf43c79424d4c9-nmls2y9k.vpce-svc-0824460b72e1a420e.us-east-1.vpce.amazonaws.com"
```

EspenAlbert replied:
Changes made in 7c55d85 ✅

Zuhairahmed commented on website/docs/r/privatelink_endpoint_service_data_federation_online_archive.html.markdown:
```diff
@@ -25,21 +28,26 @@ resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archi
   endpoint_id = "<PRIVATE-ENDPOINT-SERVICE-ID>"
```
```suggestion
  endpoint_id = "vpce-3bf78b0ddee411ba1"
```

EspenAlbert replied:
Changes made in 7c55d85 ✅

Zuhairahmed commented on website/docs/r/privatelink_endpoint_service_data_federation_online_archive.html.markdown:
```diff
@@ -25,21 +28,26 @@ resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archi
 * `provider_name` (Required) - Human-readable label that identifies the cloud service provider. 
 * `timeouts`- (Optional) The duration of time to wait for Private Endpoint Service to be created or deleted. The timeout value is definded by a signed sequence of decimal numbers with an time unit suffix such as: `1h45m`, `300s`, `10m`, .... The valid time units are:  `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`. The default timeout for Private Endpoint create & delete is `2h`. Learn more about time ...
+* `region` - (Optional) Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft- ...
+* `customer_endpoint_dns_name` - (Optional) Human-readable label to identify customer's VPC endpoint DNS name. Must follow regex: `vpce-[a-z0-9]{17}-[a-z0-9]{8}.vpce-svc-[a-z0-9]{17}.(.*).vpce.amazonaws.com`
```
```suggestion
* `customer_endpoint_dns_name` - (Optional) Human-readable label to identify customer's VPC endpoint DNS name.
```

EspenAlbert replied:
Changes made in 7c55d85 ✅

Zuhairahmed committed:
Update privatelink_endpoint_service_data_federation_online_archives.html.markdown

Zuhairahmed committed:
Update privatelink_endpoint_service_data_federation_online_archive.html.markdown

Zuhairahmed commented on website/docs/d/privatelink_endpoint_service_data_federation_online_archive.html.markdown:
```diff
@@ -46,6 +46,8 @@ In addition to all arguments above, the following attributes are exported:
 * `type` - Human-readable label that identifies the resource type associated with this private endpoint.
 * `comment` - Human-readable string to associate with this private endpoint.
 * `provider_name` - Human-readable label that identifies the cloud service provider. 
+* `region` - Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/).
```
@EspenAlbert once you are aligned can you make sure to update the 2 Data Sources files to align definitions edits made in Resource file for paramters `customer_endpoint_dns_name` and `endpoint_id`

EspenAlbert replied:
Changes made in 7c55d85 ✅

Zuhairahmed commented on website/docs/d/privatelink_endpoint_service_data_federation_online_archives.html.markdown:
```diff
@@ -22,7 +22,7 @@ resource "mongodbatlas_project" "atlas-project" {
 
 resource "mongodbatlas_privatelink_endpoint_service_data_federation_online_archive" "test" {
   project_id = mongodbatlas_project.atlas-project.id
-  endpoint_id = "<PRIVATE-ENDPOINT-SERVICE-ID>"
```
Also @EspenAlbert can you update examples in 2 Data Sources file to include `customer_endpoint_dns_name` and `endpoint_id` parameters

EspenAlbert replied:
Changes made in 7c55d85 ✅

Zuhairahmed reviewed (approved):
Left a few comments, questions, and made a few commits directly to this PR. I don't need to review again, overall great first PR @EspenAlbert!

EspenAlbert committed:
test: use an env-var for customerEndpointDNSName

Need to support testing in different environments and value should be set based on the vpc endpoint created outside the test.

EspenAlbert committed:
docs: change description of region and customer_endpoint_dns_name + update examples with actual example values instead of placeholders

EspenAlbert committed:
docs: update example for data-federation-online-archive

EspenAlbert committed:
docs: ensure example is formatted

maastha reviewed (approved):
LGTM

lantoli commented on examples/mongodbatlas_privatelink_endpoint/aws/data-federation-online-archive/atlas-privatelink.tf:
```diff
@@ -5,8 +5,10 @@ resource "mongodbatlas_privatelink_endpoint" "pe_east" {
+  project_id                 = var.project_id
+  endpoint_id                = aws_vpc_endpoint.vpce_east.id
+  provider_name              = "AWS"
+  comment                    = "Terraform Acceptance Test"
```
knitpit: i would change this comment as it's in the example folder and it's intended to be used by customers as a template, e.g.: "Example of data federation online archive"

EspenAlbert replied:
Updated in d62a3ae 👍

EspenAlbert committed:
docs: rename the comment of example
